### PR TITLE
CI: ci.py refactoring

### PR DESCRIPTION
**Pull request by maxknv**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [x]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit b4d8a078c5b2a4cefa6ca3db501f422448bb5622 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64734/b4d8a078c5b2a4cefa6ca3db501f422448bb5622/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: tests/ci/ci_cache.py [hunk 1]
```diff
@@ -0,0 +1,818 @@
...
+                                f"NOTE: Job [{job_name}:{batch}] finished failed - do not add to ready"
+                            )
+                            job_config.pending_batches.remove(batch)
+
+                        if not job_config.pending_batches:
+                            await_finished.add(job_name)
+
+                for job in await_finished:
+                    self.jobs_to_skip.append(job)
+                    del self.jobs_to_wait[job]
```
Comment (azat) [inline tests/ci/ci_cache.py:756]:
```
Traceback (most recent call last):
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/ci/ci.py", line 1238, in <module>
    sys.exit(main())
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/ci/ci.py", line 979, in main
    ci_cache.await_pending_jobs(pr_info.is_release)
  File "/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/ci/ci_cache.py", line 756, in await_pending_jobs
    del self.jobs_to_wait[job]
KeyError: 'package_msan'
```

https://github.com/ClickHouse/ClickHouse/actions/runs/9380634423/job/25828094637?pr=58349

Comment (azat) [reply tests/ci/ci_cache.py:756]: @maxknv mind taking a look why RunConfig timed out and fails eventually? https://github.com/ClickHouse/ClickHouse/pull/58349

Comment (azat) [reply tests/ci/ci_cache.py:756]: Looks like it had been fixed by https://github.com/ClickHouse/ClickHouse/pull/64825

**Pull request merged**